### PR TITLE
pkp/pkp-lib#5491 Improved the harvest performance

### PR DESCRIPTION
**Pull request by jonasraoni**

### CODE REVIEW

File: plugins/generic/mysqlIndex/SearchDAO.inc.php [hunk 1]
```diff
@@ -260,6 +266,40 @@ function insertObject($recordId, $fieldId, &$pos, $date = null, $flush = true) {
...
+	 * @param $flush boolean Whether or not to flush values of an existing object
+	 */
+	function insertObjects($recordId, $fieldIds, $date = null, $flush = true) {
+		if (!count($fieldIds)) {
+			return;
+		}
+
+		$this->update('
+			INSERT IGNORE INTO search_objects (record_id, raw_field_id, object_time)
+			VALUES ' . implode(', ', array_map(function ($fieldId) use ($recordId, $date) {
```
Comment (asmecher) [inline plugins/generic/mysqlIndex/SearchDAO.inc.php:287]: The `docs/README` requirements still specify PHP 4.2.x; mind bumping it to something less prehistoric while you're at it (e.g. PHP 5.6)? This syntax will require that.

File: plugins/generic/mysqlIndex/SearchDAO.inc.php [hunk 2]
```diff
@@ -277,6 +317,45 @@ function insertObjectKeyword($objectId, $keyword, $position) {
...
+		foreach ($fields as $fieldId => $keywords) {
+			$position = 0;
+			foreach ($keywords as $keyword) {
+				$uniqueKeywords[$keyword] = 0;
+				$objectKeywords[] = 'SELECT ' . (int) $fieldId . ' f,' . $this->_dataSource->qstr($keyword) . ' v,' . $position++ . ' p';
+			}
+		}
+		foreach (array_chunk(array_keys($uniqueKeywords), 1e3) as $keywords) {
+			$this->update('
+				INSERT IGNORE INTO search_keyword_list (keyword_text)
```
Comment (asmecher) [inline plugins/generic/mysqlIndex/SearchDAO.inc.php:340]: I don't think PostgreSQL supports `INSERT IGNORE`...

**Comment by asmecher** [comment]
Happy to see this getting some attention, @jonasraoni! I just made a few comments, almost all trivial...

File: plugins/generic/mysqlIndex/SearchDAO.inc.php (hunk 1 above)
Comment (jonasraoni) [reply plugins/generic/mysqlIndex/SearchDAO.inc.php:287]: I'm using PHP 7.1 locally and the guys are already using 7.2 in the PKP servers (which is hosting some instances of the Harvester + PKP Index), so maybe we can follow the OJS?

File: plugins/generic/mysqlIndex/SearchDAO.inc.php (hunk 2 above)
Comment (jonasraoni) [reply plugins/generic/mysqlIndex/SearchDAO.inc.php:340]: I normally would write an `INSERT...SELECT`, but this syntax was looking interesting and the fact that the plugin is living under "/mysqlIndex" gave me an extra comfort to proceed =]

I think it's ok to leave it this way, as my next proposal is to replace the manual indexing in a next iteration. But I can update if you think it's better.

Comment (asmecher) [reply plugins/generic/mysqlIndex/SearchDAO.inc.php:340]: Of course, I'd forgotten about that! Please go ahead.

File: plugins/generic/mysqlIndex/SearchDAO.inc.php (hunk 1 above)
Comment (asmecher) [reply plugins/generic/mysqlIndex/SearchDAO.inc.php:287]: That's fine; I don't think there's much benefit to maintaining PHP 5.x support.

**Comment by asmecher** [comment]
Please go ahead and merge when you're ready!

**Comment by jonasraoni** [comment]
@asmecher I've bumped the minimum versions and also updated some links/texts, let me know if it's ok... Are we going to bump the OHS version? Maybe I should move these things to /master as well?

Also, I was going to update the headers (to use GPL v3, copyright year, ...), remove the PHP's closing tag and such small cleanings, but I guess you prefer such updates divided in many PRs, right?

Then, I'm just curious about how much effort it would take to use the latest pkp-lib 🤔

**Comment by asmecher** [comment]
@jonasraoni, great, those modernizations would be good to include -- assuming you're OK with `sed` or similar!

The latest `pkp-lib` would be an interesting experiment; I suspect you'd spend quite a bit of time tinkering with the CSS, but probably little else. If you want to make an experiment please go for it, but watch out for black holes.

**Pull request merged**